File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/Callbacks/CallBackHandler.java
package kore.botssdk.audiocodes.webrtcclient.Callbacks;


import android.content.Context;

import com.audiocodes.mv.webrtcsdk.im.InstanceMessageStatus;
import com.audiocodes.mv.webrtcsdk.session.CallState;

import java.util.ArrayList;
import java.util.List;

public class CallBackHandler {
    private static final List<LoginStateChanged> loginStateChangeCallBacks = new ArrayList<>();
    private static final List<CallStateChanged> callStateChangeCallBacks = new ArrayList<>();
    private static final List<ChatCallback> chatCallBacks = new ArrayList<>();
    private static final List<TabChangeCallback> tabChangeCallbacks = new ArrayList<>();

    //LoginStateChanged
    public interface LoginStateChanged {
        public void loginStateChange(Context context, boolean state);
    }

    public static void loginStateChange(Context context, boolean state) {
        for (int i = 0; i < loginStateChangeCallBacks.size(); ++i) {
            loginStateChangeCallBacks.get(i).loginStateChange(context, state);
        }
    }

    public static void unregisterLoginStateChange(LoginStateChanged cb) {
        loginStateChangeCallBacks.remove(cb);
    }

    public static void registerLoginStateChange(LoginStateChanged cb) {
        loginStateChangeCallBacks.add(cb);
    }

    //CallStateChanged
    public interface CallStateChanged {
        public void callStateChanged(CallState callState);
    }

    public static void callStateChanged(CallState callState) {
        for (int i = 0; i < callStateChangeCallBacks.size(); ++i) {
            callStateChangeCallBacks.get(i).callStateChanged(callState);
        }
    }

    public static void unregisterCallStateChanged(CallStateChanged cb) {
        callStateChangeCallBacks.remove(cb);
    }

    public static void registerCallStateChanged(CallStateChanged cb) {
        callStateChangeCallBacks.clear();
        callStateChangeCallBacks.add(cb);
    }

    //ChatCallback
    public interface ChatCallback {
        void onNewMessage(String user, String message);

        void onMessageStatus(InstanceMessageStatus instanceMessageStatus, long ID);
    }

    public static void onNewMessage(String user, String message) {
        for (int i = 0; i < chatCallBacks.size(); ++i) {
            chatCallBacks.get(i).onNewMessage(user, message);
        }
    }

    public static void onMessageStatus(InstanceMessageStatus instanceMessageStatus, long ID) {
        for (int i = 0; i < chatCallBacks.size(); ++i) {
            chatCallBacks.get(i).onMessageStatus(instanceMessageStatus, ID);
        }
    }

    public static void unregisterChatCallback(ChatCallback chatCallback) {
        chatCallBacks.remove(chatCallback);
    }

    public static void registerChatCallback(ChatCallback chatCallback) {
        chatCallBacks.clear();
        chatCallBacks.add(chatCallback);
    }

    //TabChangeCallback
    public interface TabChangeCallback {
        void onTabChange(int tabIndex);
    }

    public static void onTabChange(int tabIndex) {
        for (int i = 0; i < tabChangeCallbacks.size(); ++i) {
            tabChangeCallbacks.get(i).onTabChange(tabIndex);
        }
    }

    public static void unregisterTabChangeCallback(TabChangeCallback tabChangeCallback) {
        chatCallBacks.remove(tabChangeCallback);
    }

    public static void registerTabChangeCallback(TabChangeCallback tabChangeCallback) {
        tabChangeCallbacks.clear();
        tabChangeCallbacks.add(tabChangeCallback);
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/callbacks/CallBackHandler.kt
package com.kore.ui.audiocodes.webrtcclient.callbacks

import android.annotation.SuppressLint
import com.audiocodes.mv.webrtcsdk.im.InstanceMessageStatus
import com.audiocodes.mv.webrtcsdk.session.CallState

object CallBackHandler {
    private val loginStateChangeCallBacks: MutableList<LoginStateChanged> = ArrayList()
    private val callStateChangeCallBacks: MutableList<CallStateChanged> = ArrayList()
    private val chatCallBacks: MutableList<ChatCallback> = ArrayList()
    private val tabChangeCallbacks: MutableList<TabChangeCallback> = ArrayList()

    @JvmStatic
    fun loginStateChange(state: Boolean) {
        for (i in loginStateChangeCallBacks.indices) {
            loginStateChangeCallBacks[i].loginStateChange(state)
        }
    }

    @JvmStatic
    fun unregisterLoginStateChange(cb: LoginStateChanged) {
        loginStateChangeCallBacks.remove(cb)
    }

    @JvmStatic
    fun registerLginStateChange(cb: LoginStateChanged) {
        loginStateChangeCallBacks.add(cb)
    }

    @SuppressLint("UnknownNullness")
    fun callStateChanged(callState: CallState) {
        for (i in callStateChangeCallBacks.indices) {
            callStateChangeCallBacks[i].callStateChanged(callState)
        }
    }

    fun unregisterCallStateChanged(cb: CallStateChanged) {
        callStateChangeCallBacks.remove(cb)
    }

    fun registerCallStateChanged(cb: CallStateChanged) {
        callStateChangeCallBacks.clear()
        callStateChangeCallBacks.add(cb)
    }

    @JvmStatic
    fun onNewMessage(user: String, message: String) {
        for (i in chatCallBacks.indices) {
            chatCallBacks[i].onNewMessage(user, message)
        }
    }

    @JvmStatic
    @SuppressLint("UnknownNullness")
    fun onMessageStatus(instanceMessageStatus: InstanceMessageStatus, ID: Long) {
        for (i in chatCallBacks.indices) {
            chatCallBacks[i].onMessageStatus(instanceMessageStatus, ID)
        }
    }

    @JvmStatic
    @SuppressLint("UnknownNullness")
    fun unregisterChatCallback(chatCallback: ChatCallback) {
        chatCallBacks.remove(chatCallback)
    }

    @JvmStatic
    @SuppressLint("UnknownNullness")
    fun registerChatCallback(chatCallback: ChatCallback) {
        chatCallBacks.clear()
        chatCallBacks.add(chatCallback)
    }

    @JvmStatic
    fun onTabChange(tabIndex: Int) {
        for (i in tabChangeCallbacks.indices) {
            tabChangeCallbacks[i].onTabChange(tabIndex)
        }
    }

    @SuppressLint("UnknownNullness")
    fun unregisterTabChangeCallback(tabChangeCallback: TabChangeCallback): Boolean {
        return chatCallBacks.remove(tabChangeCallback as ChatCallback)
    }

    @SuppressLint("UnknownNullness")
    fun registerTabChangeCallback(tabChangeCallback: TabChangeCallback) {
        tabChangeCallbacks.clear()
        tabChangeCallbacks.add(tabChangeCallback)
    }

    //LoginStateChanged
    fun interface LoginStateChanged {
        fun loginStateChange(state: Boolean)
    }

    //CallStateChanged
    interface CallStateChanged {
        fun callStateChanged(callState: CallState)
    }

    //ChatCallback
    interface ChatCallback {
        fun onNewMessage(user: String, message: String)
        fun onMessageStatus(instanceMessageStatus: InstanceMessageStatus, ID: Long)
    }

    //TabChangeCallback
    fun interface TabChangeCallback {
        fun onTabChange(tabIndex: Int)
    }
}
